src/functional/ffront/type_info.py

```diff
@@ -165,31 +166,123 @@ def is_concretizable(symbol_type: ct.SymbolType, to_type: ct.SymbolType) -> bool
+     ...
+    functional.common.GTTypeError: Dimensions can not be promoted. Could not determine order of the following dimensions: J, K.
+    """
+    if all(isinstance(type_, ct.ScalarType) for type_ in types):
+        if not all(type_ == types[0] for type_ in types):
+            raise GTTypeError("Could not promote scalars of different dtype not implemented.")
+        if not all(type_.shape is None for type_ in types):  # type: ignore[union-attr]
+            raise NotImplementedError("Shape promotion not implemented.")
+        return types[0]
+    elif all(isinstance(type_, (ct.ScalarType, ct.FieldType)) for type_ in types):
+        dims = promote_dims(*(extract_dims(type_) for type_ in types))
+        dtype = cast(ct.ScalarType, promote(*(extract_dtype(type_) for type_ in types)))
+
+        return ct.FieldType(dims=dims, dtype=dtype)
+    raise TypeError("Expected a FieldType or ScalarType.")
+
+
+def promote_dims(
+    *dims_list: list[Dimension] | EllipsisType,
+) -> list[Dimension] | EllipsisType:
     """
-    Check if `symbol_type` has no fixed dimensionality and can be dimensionally promoted.
+    Find a unique ordering of multiple (individually ordered) lists of dimensions.
 
-    This is not to be mistaken for broadcasting, a more general concept, which allows interactions
-    between differing sets of dimensions.
+    The resulting list of dimensions contains all dimensions of the arguments
+    in the order they originally appear. If no unique order exists or a
+    contradicting order is found an exception is raised.
+
+    A modified version (ensuring uniqueness of the order) of
+    `Kahn's algirithm <https://en.wikipedia.org/wiki/Topological_sorting#Kahn's_algorithm>`_
+    is used to topologically sort the arguments.
+
+    >>> I, J, K = (Dimension(value=dim) for dim in ["I", "J", "K"])
+    >>> promote_dims([I, J], [I, J, K]) == [I, J, K]
+    True
+    >>> promote_dims([I, J], [K]) # doctest: +ELLIPSIS
+    Traceback (most recent call last):
+     ...
+    functional.common.GTTypeError: Dimensions can not be promoted. Could not determine order of the following dimensions: J, K.
+    >>> promote_dims([I, J], [J, I]) # doctest: +ELLIPSIS
+    Traceback (most recent call last):
+     ...
+    functional.common.GTTypeError: Dimensions can not be promoted. The following dimensions appear in contradicting order: I, J.
     """
-    # scalars can always be broadcasted
-    if type_class(symbol_type) is ct.ScalarType:
-        return True
-    # symbol_type must be either zero- or any dimensional or have same dimensionality as to_type
-    elif type_class(to_type) is ct.FieldType:
-        dims = extract_dims(symbol_type)
-        return dims in [[], Ellipsis] or _is_sublist(dims, extract_dims(to_type))
-    return False
+    # build a graph with the vertices being dimensions and edges representing
+    #  the order between two dimensions
+    graph: dict[Dimension, set[Dimension]] = {}
```

Comment:
It looks like`graph[node]` is encoding the _predecessors_ of `node` instead of its _sucessors_, which somehow feels more natural to me and then forces the algorithm to build the sorted list going backwards. I think it'd be simpler to understand working forwards but, in any case, I strongly suggest to rename this dict to something more explicit like `predecessors` (current implementation) or `sucessors` (my preferred solution).

Reply:
Indeed this is somewhat strange. I initially wanted to use the topological sorting facilities of the Python standard library (https://docs.python.org/3/library/graphlib.html), which use this format (and implemented the algorithm myself to ensure the uniqueness property). I would like to keep it "compatible" with how graphlib does it, but made this more explicit in the code now (using comments and renaming of some variables). What do you think?